enhancements/cluster-capabilities-api.md

```diff
@@ -0,0 +1,328 @@
+  in the kubelet? How does an n-2 kubelet without this feature available behave
+  when this feature is used?
+- Will any other components on the node change? For example, changes to CSI, CRI
+  or CNI may require updating that component before the kubelet.
+
+## Implementation History
+
+Major milestones in the life cycle of a proposal should be tracked in `Implementation
+History`.
+
+## Drawbacks
+
+The idea is to find the best form of an argument why this enhancement should _not_ be implemented.
+
+## Alternatives
+
+### Add fields to existing configuration APIs
+
+We could add the fields we know about today to the
+`infrastructure.config.openshift.io` API, instead of creating a new
+API. This would be less work, initially, but muddies the meaning of
+the APIs.
+
+### Expose the cluster profile name directly
+
+We could add a `clusterProfile` field to an existing or new
+configuration API. This would require each operator to interpret each
+new profile, which would make maintaining profiles even more work than
+it is now. It may also encourage us to create more profiles, when we
+have stated elsewhere that our goal is to have as few as possible.
+
+### Use the cluster profile to have different Deployments for operators
+
+We could use different manifests to have the cluster-version-operator
+deploy operators with different configuration (via command line
+arguments or environment variables). This would give us a way to
+configure operators differently for each profile, but those settings
+would be less discoverable and we would not be able to offer users a
+way to change them at runtime (if we ever want to do that).
+
+### Recognize HA/non-HA configuration based on control plane node count
+
+Operators could look at the number of control plane nodes and use that
+information to determine whether HA is enabled (3 or more nodes) or
+not (1 node). This would require giving the operator read access to
+Node resources, and would be subject to race conditions as the control
+plane forms. For example, the cluster-etcd-operator [requires 3
+control plane
+nodes](https://github.com/openshift/cluster-etcd-operator/blob/98590e6ecfe282735c4eff01432ae40b29f81202/pkg/etcdenvvar/etcd_env.go#L72)
+to exist before deploying its operand. If it counted nodes, it could
+not tell the difference between a cluster starting to form with 1
+node, then more, or a cluster that is only ever going to include 1
+node.
+
+### Use operator-specific APIs to manage the behaviors
+
+The cluster-etcd-operator and cluster-authentication-operator both
+include flags to enter an unsupported non-HA operating mode. We could
+change those APIs to make the flags supported, and add similar flags
+to other components, as needed. This would offer more flexibility,
+since each component could be configured separately, but would make
+integrating with the installer and cluster profiles much more
+complicated since each component would need to be configured
+separately.
```

Comment:
Do we have a realistic number of component-specific knobs?  So far I've heard about etcd and auth knobs for HA-ness, and monitoring knobs for resource consumption.  Three knobs is not "much more complicated" than a central knob in my subjective opinion.  It's certainly possible that there are other component-specific knobs out there today, or planned for the future if we don't have these cluster-scoped knobs, but I think it's worth digging into that a bit more to make a stronger case for cluster-scoped knobs.  And as you point out with the precedence comments in the *Proposal* section, having cluster-scoped knobs still leaves space for component-scoped knobs, for components that want independent flexibility.

Reply:
The implementation details section of #504 mentions cluster-ingress-operator and machine-config-operator as needing to honor the API. Neither has their own API for these behaviors today.

Reply:
I added the operators we know will be affected to this section of the doc.